### PR TITLE
Base client

### PR DESCRIPTION
**Pull request by shellbj**

### CODE REVIEW

File: core/pom.xml [hunk 1]
```diff
@@ -42,6 +48,19 @@
             <artifactId>mockito-core</artifactId>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>io.opentracing</groupId>
+            <artifactId>opentracing-mock</artifactId>
+            <scope>test</scope>
```
Comment (jeffbakerexpedia) [inline core/pom.xml]: Is there a reason that you don't specify versions of your dependencies?

File: core/src/main/java/com/expedia/www/haystack/client/LogData.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    private final Long timestamp;
+    private final Map<String, ?> fields;
+    private final String message;
+    private final Object payload;
+
+    public LogData(Long timestamp, String message) {
+        this.timestamp = timestamp;
+        this.message = message;
+        this.payload = null;
+        this.fields = null;
```
Comment (jeffbakerexpedia) [inline core/src/main/java/com/expedia/www/haystack/client/LogData.java]: I personally prefer a single constructor that sets all instance fields (four in your case) with overloaded constructors with the different signatures. Less code, but YMMV so if you like it better setting all fields in each constructor that's OK.

File: core/src/main/java/com/expedia/www/haystack/client/Span.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+	 */
+    protected synchronized void finishedCheck(String format, Object... args) {
+        if (finished.get()) {
+            RuntimeException e = new IllegalStateException(String.format(format, args));
+            errors.add(e);
+            throw e;
+        }
+    }
+
+	/**
```
Comment (jeffbakerexpedia) [inline core/src/main/java/com/expedia/www/haystack/client/Span.java]: Tabs mixed with spaces? I have IntelliJ configured to turn tabs into spaces; makes it easier to read regardless of the reader software.

File: core/src/main/java/com/expedia/www/haystack/client/Span.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+          return this;
+      }
+      finishedCheck("Setting a log event (%s:%s) on a finished span", timestampMicroseconds, fields); 
+      logs.add(new LogData(timestampMicroseconds, fields));
+      return this;
+	}
+
+	@Override
+	public Span log(Map<String, ?> fields) {
+      return log(System.nanoTime(), fields);
```
Comment (jeffbakerexpedia) [inline core/src/main/java/com/expedia/www/haystack/client/Span.java]: Isn't this method passing nanoseconds to a method that expects microseconds?

File: core/src/main/java/com/expedia/www/haystack/client/Span.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+          return this;
+      }
+      finishedCheck("Setting a log event (%s:%s:%s) on a finished span", timestampMicroseconds, eventName, payload); 
+      logs.add(new LogData(timestampMicroseconds, eventName, payload));
+      return this;
+	}
+
+	@Override
+	public Span log(String eventName, Object payload) {
+      return log(System.nanoTime(), eventName, payload);
```
Comment (jeffbakerexpedia) [inline core/src/main/java/com/expedia/www/haystack/client/Span.java]: Same comment about time units. If I see it again, I won't mention it.

File: core/src/main/java/com/expedia/www/haystack/client/Tracer.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+                                 UUID.randomUUID(),
+                                 parent.getContext().getSpanId(),
+                                 baggage);
+      }
+
+      private SpanContext createContext() {
+          if (references.isEmpty()) {
+              return createNewContext();
+          }
+          return createDependantContext();
```
Comment (jeffbakerexpedia) [inline core/src/main/java/com/expedia/www/haystack/client/Tracer.java]: dependent (emphasis on the third 'e'). Yeah, I won the junior high spelling bee.

File: core/pom.xml (hunk 1 above)
Comment (shellbj) [reply core/pom.xml]: It is; see the parent pom scope under `dependencyManagement`.  If I missed one that that would be a different story.

File: core/src/main/java/com/expedia/www/haystack/client/LogData.java (hunk 2 above)
Comment (shellbj) [reply core/src/main/java/com/expedia/www/haystack/client/LogData.java]: All four aren't actually able to be set at once.  I will refactor one of them to remove some of the duplication.

File: core/src/main/java/com/expedia/www/haystack/client/Span.java (hunk 3 above)
Comment (shellbj) [reply core/src/main/java/com/expedia/www/haystack/client/Span.java]: Blame eclipse for being stupid... I'll clean it up.

**Pull request merged**